stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -670,13 +670,31 @@ function opnorm1(A::AbstractMatrix{T}) where T
     return convert(Tnorm, nrm)
 end
 
+# Uses power iteration to compute the maximal singular value.
+# falls back to svdvals if it runs into numerics issues.
 function opnorm2(A::AbstractMatrix{T}) where T
     require_one_based_indexing(A)
     m,n = size(A)
     Tnorm = typeof(float(real(zero(T))))
     if m == 0 || n == 0 return zero(Tnorm) end
     if m == 1 || n == 1 return norm2(A) end
-    return svdvals(A)[1]
+    # to minimize the chance of x being orthogonal to the largest eigenvector
+    x = randn(Tnorm, n)
+    tmp = zeros(Tnorm, m)
+    At = A'
+    v = one(Tnorm)
+    # this will converge quickly as long as the top two eigenvalues are distinct
+    for i in 1:n
+        mul!(x, At, mul!(tmp, A, x))
+        newv = norm(x)
+        # the numerics got very wonky
+        !isfinite(newv) && return first(svdvals(A))
```

Comment:
You're potentially going to hit this case as soon as `A` has entries `> sqrt(floatmax())`.   Seems like you could avoid it by doing:
```
tmpnorm = norm(mul!(tmp, A, x))
iszero(tmpnorm) && return tmpnorm
tmp .*= inv(tmpnorm)
xnorm = norm(mul!(x, At, tmp))
newv = sqrt(xnorm) * sqrt(tmporm)
```
(and then return `newv`, not its sqrt)

Not only would this eliminate spurious overflow, but I think you could then just do `
!isfinite(newv) && return newv` since any case with `!isfinite(newv)` would then also have non-finite svdvals (e.g. `A` has `Inf` or `NaN` entries).

Reply:
It might be that `for i in 1:n` needs some work?

```julia
julia> opnorm2(complex.(randn(2,2),randn(2,2)))
ERROR: InexactError: Float64(1.894212444018611 - 0.6331930071371694im)
[...]
```